### PR TITLE
bpo-31337 Close a minor NULL dereference opportunity

### PR DESCRIPTION
**Pull request by warsaw**
https://bugs.python.org/issue31337

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -3360,6 +3360,10 @@ compiler_subdict(struct compiler *c, expr_ty e, Py_ssize_t begin, Py_ssize_t end
         }
         for (i = begin; i < end; i++) {
             key = get_const_value((expr_ty)asdl_seq_GET(e->v.Dict.keys, i));
+            if (key == NULL) {
+                Py_DECREF(keys);
+                return 0;
```
Comment (serhiy-storchaka) [inline Python/compile.c]: In this case an error is returned, but no exception is set. This looks like a bug to me. We have fixed a number of similar bugs recent times (thanks to Victor's runtime checks).

Comment (warsaw) [reply Python/compile.c]: @serhiy-storchaka Thanks!  Where would I see an example of Victor's runtime checks?  I didn't notice anything in local tests.

What exception do you think we should raise?  `MemoryError`, `RuntimeError` or something else?

Comment (warsaw) [reply Python/compile.c]: `PyErr_BadInternalCall()` is the most obvious.

**Comment by serhiy-storchaka** [comment]
> @serhiy-storchaka Thanks!  Where would I see an example of Victor's runtime
> checks?  I didn't notice anything in local tests.

See for example lines 3566-3573 in ceval.c. And there should be similar code 
in several other places.

> What exception do you think we should raise?  `MemoryError`, `RuntimeError`
> or something else?

Note that this is a dead code! I think we shouldn't add it at all.

**Comment by warsaw** [comment]
@serhiy-storchaka Oh, I see what your reference to "Victor's runtime checks" are.  Yes, I knew about the `!PyErr_Occurred()` check, but not the reference.

I don't agree that this is *obviously* dead code.  You have to reason about the control flow, and assume that `-DNDEBUG` was not given, and that nothing changes to subtle break the assumptions in this code.  With additional feedback, I am however willing to forego applying this in favor of a more comprehensive fix for bpo-31338.

**Comment by warsaw** [comment]
Closing this in favor of bpo-33138

**Pull request closed without merging**